com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/persistence/query/impl/AbstractQueryImpl.java

```diff
@@ -0,0 +1,110 @@
+    private String orderAttribute;
+
+    public AbstractQueryImpl(EntityManager entityManager) {
+        this.entityManager = entityManager;
+        this.criteriaBuilder = entityManager.getCriteriaBuilder();
+    }
+
+    @Override
+    public T limitOnSelect(int limit) {
+        this.limit = limit;
+        return getSelf();
+    }
+
+    @Override
+    public T offsetOnSelect(int offset) {
+        this.offset = offset;
+        return getSelf();
+    }
+
+    protected T setOrder(String orderAttribute, OrderDirection orderDirection) {
+        this.orderAttribute = orderAttribute;
+        this.orderDirection = orderDirection;
+        return getSelf();
+    }
+
+    private <E> void applyLimitAndOffset(TypedQuery<E> typedQuery) {
+        if (limit != null) {
+            typedQuery.setMaxResults(limit);
+        }
+        if (offset != null) {
+            typedQuery.setFirstResult(offset);
+        }
+    }
+
+    private <E> CriteriaQuery<E> applyOrder(CriteriaQuery<E> criteriaQuery, Root<E> root) {
+        if (orderAttribute == null) {
+            return criteriaQuery;
+        }
+        if (orderDirection == OrderDirection.ASCENDING) {
+            return criteriaQuery.orderBy(criteriaBuilder.asc(root.get(orderAttribute)));
+        }
+        return criteriaQuery.orderBy(criteriaBuilder.desc(root.get(orderAttribute)));
+    }
+
+    protected <E> TypedQuery<E> createQuery(EntityManager entityManager, QueryCriteria criteria, Class<E> dtoClass) {
+        CriteriaQuery<E> criteriaQuery = criteriaBuilder.createQuery(dtoClass);
+        Root<E> root = criteriaQuery.from(dtoClass);
+        criteriaQuery.where(criteria.toQueryPredicates(root)
+                                    .toArray(new Predicate[0]));
+        criteriaQuery = applyOrder(criteriaQuery, root);
+        TypedQuery<E> typedQuery = entityManager.createQuery(criteriaQuery);
+        applyLimitAndOffset(typedQuery);
+        return typedQuery;
+    }
+
+    protected <E> javax.persistence.Query createDeleteQuery(EntityManager entityManager, QueryCriteria criteria, Class<E> dtoClass) {
+        CriteriaDelete<E> deleteQuery = criteriaBuilder.createCriteriaDelete(dtoClass);
+        Root<E> root = deleteQuery.from(dtoClass);
+        deleteQuery.where(criteria.toQueryPredicates(root)
+                                  .toArray(new Predicate[0]));
+        return entityManager.createQuery(deleteQuery);
+    }
+
+    protected <E> E executeInTransaction(Function<EntityManager, E> function) {
+        return new TransactionalExecutor<E>(entityManager).execute(function);
+    }
+
+    @Override
+    public R singleResultOrNull() {
+        try {
+            return singleResult();
+        } catch (NoResultException e) {
+            return null;
+        }
+    }
+
+    protected CriteriaBuilder getCriteriaBuilder() {
+        return criteriaBuilder;
+    }
+
+    @SuppressWarnings("unchecked")
+    public T getSelf() {
+        return (T) this;
```

Comment:
It is mandatory from the usage of a recursive type bound (check out the class definition), otherwise it would not compile. Another possibility is to return Query<R, T> instead of T, but that would break the chain for the concrete implementations (OperationQuery, Conf.. etc).
I double checked this.